R/resource.R

```diff
@@ -0,0 +1,160 @@
+##'   concurrently run on the node while it runs your task.
+##'
+##' @param runtime Set this to specify a time limit for running your job.
+##'   Acceptable formats are either an integer number of minutes, or
+##'   strings in the form `"h:m"`, or `"d:h:m"` to specify
+##'   days, hours and minutes.
+##'
+##' @param hold_until Specify your task should wait in the queue until
+##'   a certain time, or for a certain period. For the former, this can
+##'   be a POSIXt, a Date, the special strings "tonight" (7pm), "midnight",
+##'   or "weekend" (midnight Saturday morning). To delay for a period,
+##'   this can be an integer number of minutes, or a string in the form
+##'   `"h:m"` for hours and minutes.
+##'
+##' @param memory_per_node Specify your task can only run on a node
+##'   with at least the specified memory. This is an integer,
+##'   measured in gigabytes, or strings such as `"64Gb"` or `"1Tb"`.
+##'
+##' @param memory_per_process If you can provide an estimate of how
+##'   much RAM your task requires, then the cluster can ensure the
+##'   total memory required by running multiple tasks on a node
+##'   does not exceed how much memory the node has. Specify this as
+##'   an integer number of gigabytes, or characters such as `"10Gb"`
+##'
+##' @param requested_nodes If you have been in touch with us or DIDE
+##'   IT, and you need to run your task on a selection of named compute
+##'   nodes, then specify this here as a vector of strings for the node names.
+##'
+##' @param priority If the tasks you are launching are low priority, you can
+##'   allow other queuing tasks to jump over them, by setting the priority to
+##'   to `low`; otherwise, the default is `normal`. These are the only
+##'   acceptable values.
+##'
+##' @param queue Specify a particular queue to submit your tasks to.
+##'   This is in development as we decide over time what
+##'   queues we best need for DIDE's common workflows. See the Details
+##'   for more information
+##'
+##' @return A hipercow_resource list, which can be provided when you submit
+##'   tasks.
+##'
+##' @inheritParams hipercow_configure
+##'
+##' @export
+
+task_resources <- function(cores = 1L,
+                           exclusive = FALSE,
+                           runtime = NULL,
+                           hold_until = NULL,
+                           memory_per_node = NULL,
+                           memory_per_process = NULL,
+                           requested_nodes = NULL,
+                           priority = NULL,
+                           queue = NULL,
+                           driver = NULL,
+                           root = NULL) {
+  root <- hipercow_root(root)
+  dat <- hipercow_driver_prepare(driver, root, rlang::current_env())
+
+  validate_cores(cores)
+  if (rlang::is_integerish(cores)) {
+    cores <- as.integer(cores)
+  }
+
+  assert_scalar_logical(exclusive)
+  is.null(runtime) || validate_runtime(runtime)
```

Comment:
```suggestion
  ret <- list(
    runtime = validate_runtime(runtime),
    ...) # etc
```

then in `validate_runtime` include the null bit:

```
validate_runtime <- function(runtime) {
  if (is.null(runtime)) {
    return(NULL) # happy path, early exit, equivalent to what you had before
  }
  assert_scalar(runtime) # can also pass `call` through here for nice traces
  if (is.integer(runtime)) {
    # check positive etc
  } else if (is.character(runtime)) {
    stop("not yet implemented")
  } else {
    cli::cli_abort("Invalid value for '{runtime}'; expected a NULL, integer or string")
  }
  runtime
}
```

and similarly for the rest of these functions

Reply:
Yeah - I started doing this - it was feeling quite sprawly, but is the right way...